contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,557 @@
+through the gRPC apis.
+
+# End User story
+
+When setting up the cluster the admin knows what kind of devices are present
+on the different machines and therefore can select what devices he want to
+enable.
+
+The cluster admin knows his cluster has NVIDIA GPUs therefore he deploys
+the NVIDIA device plugin through:
+`kubectl create -f nvidia.io/device-plugin.yml`
+
+The device plugin lands on all the nodes of the cluster and if it detects that
+there are no GPUs it terminates. However, when there are GPUs it reports them
+to Kubelet and starts it's gRPC server to monitor devices and hook into the
+container creation process.
+
+Device Plugins reporting non-GPU Devices are advertised as Extended resources
+of the shape `extensions.kubernetes.io/vendor-device` GPUs are advertised
+as `nvidia-gpu`.
+
+Devices can be selected using the same process as for OIRs in the pod spec.
+
+1. A user submits a pod spec requesting X GPUs (or devices) through OIR
+2. The scheduler filters the nodes which do not match the resource requests
+3. The pod lands on the node and Kubelet decides which device
+   should be assigned to the pod
+4. Kubelet calls `Allocate` on the matching Device Plugins
+5. The user deletes the pod or the pod terminates
+
+When receiving a pod which requests Devices kubelet is in charge of:
+  * deciding which device to assign to the pod's containers 
+  * Calling the `Allocate` function with the list of devices
+
+The scheduler is still in charge of filtering the nodes which cannot
+satisfy the resource requests.
+
+# Device Plugin
+
+## Introduction
+
+The device plugin is structured in 3 parts:
+1. Registration: The device plugin advertises it's presence to Kubelet
+2. ListAndWatch: The device plugin advertises a list of Devices to Kubelet
+   and sends it again if the state of a Device changes
+3. Allocate: When creating containers, Kubelet calls the device plugin's
+   `Allocate` function so that it can run device specific instructions (gpu
+    cleanup, QRNG initialization, ...) and instruct Kubelet how to make the
+    device available in the container.
+
+## Registration
+
+When starting the device plugin is expected to make a (client) gRPC call
+to the `Register` function that Kubelet exposes.
+
+The communication between Kubelet is expected to happen only through Unix
+sockets and follow this simple pattern:
+1. The device plugins sends a `RegisterRequest` to Kubelet (through a
+   gRPC request)
+2. Kubelet answers to the `RegisterRequest` with a `RegisterResponse`
+   containing any error Kubelet might have encountered
+3. The device plugin start it's gRPC server if it did not recieve an
+   error
+
+Note that in the register request, the device plugin lists the labels
+it wants the node to advertise.
+If the labels are conflicting with pre-existing labels, Kubelet does
+not answer with a successfull `RegisterRequest`.
+
+In the current alpha version, labels will be used as properties for
+the device.
```

Comment:
Suggest we leave the label discussion out in this document. I think we will need a separate PR to discuss device property exporting and selection.